implementation/src/main/java/io/smallrye/mutiny/operators/UniCache.java

```diff
@@ -1,139 +1,181 @@
-    private static final int SUBSCRIBING = 1;
-    private static final int SUBSCRIBED = 2;
-    private static final int COMPLETED = 3;
-    private final AtomicReference<UniSubscription> subscription = new AtomicReference<>();
-    private final List<UniSubscriber<? super I>> subscribers = new ArrayList<>();
-    private int state = 0;
-    private I item;
-    private Throwable failure;
+    private enum State {
+        INIT,
+        SUBSCRIBING,
+        SUBSCRIBED,
+        CACHING
+    }
+
+    private final BooleanSupplier invalidationRequested;
+
+    private final AtomicReference<State> state = new AtomicReference<>(State.INIT);
+
+    private final AtomicInteger wip = new AtomicInteger();
+    private final List<UniSerializedSubscriber<? super I>> awaitingSubscription = synchronizedList(new ArrayList<>());
+    private final List<UniSerializedSubscriber<? super I>> awaitingResult = synchronizedList(new ArrayList<>());
+
+    private volatile UniSubscription upstreamSubscription;
+    private volatile I item;
+    private volatile Throwable failure;
 
     UniCache(Uni<? extends I> upstream) {
+        this(upstream, () -> false);
+    }
+
+    UniCache(Uni<? extends I> upstream, BooleanSupplier invalidationRequested) {
         super(nonNull(upstream, "upstream"));
+        this.invalidationRequested = invalidationRequested;
     }
 
     @Override
     protected void subscribing(UniSerializedSubscriber<? super I> subscriber) {
-        Runnable action = null;
-        synchronized (this) {
-            switch (state) {
-                case NOT_INITIALIZED:
-                    // First subscriber,
-                    state = SUBSCRIBING;
-                    action = () -> AbstractUni.subscribe(upstream(), this);
-                    subscribers.add(subscriber);
-                    break;
-                case SUBSCRIBING:
-                    // Subscription pending
-                    subscribers.add(subscriber);
-                    break;
-                case SUBSCRIBED:
-                    // No item yet, but we can provide a Subscription
-                    subscribers.add(subscriber);
-                    action = () -> subscriber.onSubscribe(() -> onCancellation(subscriber));
-                    break;
-                case COMPLETED:
-                    // Result already computed
-                    subscribers.add(subscriber);
-                    action = () -> {
-                        // We must first pass a subscription
-                        subscriber.onSubscribe(() -> onCancellation(subscriber));
-                        replay(subscriber);
-                    };
-                    break;
-                default:
-                    throw new IllegalStateException("Unknown state: " + state);
+        awaitingSubscription.add(subscriber);
+        if (invalidationRequested.getAsBoolean() && state.get() != State.SUBSCRIBING) {
+            state.set(State.INIT);
```

Comment:
couldn't there be races in the `if` condition?

Reply:
I suppose that there is no causality between `invalidationRequested` and `state`.